x/liquidgov/keeper/hooks.go

```diff
@@ -0,0 +1,131 @@
+		//	k.CompleteUnlocking(ctx, denom, owner) // returns mature lockups to owners and deletes lockups
+	}
+}
+
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochInfo epochstypes.EpochInfo) {}
+
+// Hooks wrapper struct for incentives keeper
+type Hooks struct {
+	k Keeper
+}
+
+var _ epochstypes.EpochHooks = Hooks{}
+
+func (k Keeper) Hooks() Hooks {
+	return Hooks{k}
+}
+
+// epochs hooks
+func (h Hooks) BeforeEpochStart(ctx sdk.Context, epochInfo epochstypes.EpochInfo) {
+	h.k.BeforeEpochStart(ctx, epochInfo)
+}
+
+func (h Hooks) AfterEpochEnd(ctx sdk.Context, epochInfo epochstypes.EpochInfo) {
+	h.k.AfterEpochEnd(ctx, epochInfo)
+}
+
+// // Update the epoch information in the stakeibc epoch tracker
+// func (k Keeper) UpdateEpochTracker(ctx sdk.Context, epochInfo epochstypes.EpochInfo) (epochNumber uint64, err error) {
+// 	epochNumber, err = cast.ToUint64E(epochInfo.CurrentEpoch)
+// 	if err != nil {
+// 		k.Logger(ctx).Error(fmt.Sprintf("Could not convert epoch number to uint64: %v", err))
+// 		return 0, err
+// 	}
+// 	epochDurationNano, err := cast.ToUint64E(epochInfo.Duration.Nanoseconds())
+// 	if err != nil {
+// 		k.Logger(ctx).Error(fmt.Sprintf("Could not convert epoch duration to uint64: %v", err))
+// 		return 0, err
+// 	}
+// 	nextEpochStartTime, err := cast.ToUint64E(epochInfo.CurrentEpochStartTime.Add(epochInfo.Duration).UnixNano())
+// 	if err != nil {
+// 		k.Logger(ctx).Error(fmt.Sprintf("Could not convert epoch duration to uint64: %v", err))
+// 		return 0, err
+// 	}
+// 	epochTracker := stakeibctypes.EpochTracker{
+// 		EpochIdentifier:    epochInfo.Identifier,
+// 		EpochNumber:        epochNumber,
+// 		Duration:           epochDurationNano,
+// 		NextEpochStartTime: nextEpochStartTime,
+// 	}
+// 	k.SetEpochTracker(ctx, epochTracker)
+
+// 	return epochNumber, nil
+// }
+
+func (k Keeper) UpdateProposals(ctx sdk.Context) {
+	// for _, hostZone := range k.GetAllHostZone(ctx) {
+	// 	// ICQ proposals
+	// 	k.MirrorProposals(ctx, hostzone)
+	// 	//
+	// }
+	hostZone, found := k.stakeibcKeeper.GetHostZone(ctx, "GAIA") // TODO remove hardcode
+	if !found {
+		k.Logger(ctx).Info(fmt.Sprintf("hostzone not found %s...", "GAIA"))
+	} else {
+		k.MirrorProposals(ctx, hostZone)
+	}
+}
+
+// func (k Keeper) CastVotes(ctx, epochInfo epochstypes.EpochInfo) {
+// 	// get param for cast window - may be hostzone specific?
+// 	window := k.GetCastWindowParam(ctx)
+// 	// for every proposal in each host zone check for cast window and cast votes
+// 	for _, hostZone := range k.GetAllHostZone(ctx) {
+// 		for _, proposal := range k.GetProposalsForHostZone(ctx, hostZone) {
+// 			// if proposal in cast window
+// 			if epochInfo.CurrentEpochStartTime + window > proposal.votingEndTime {
+// 				// tally votes
+// 				vote := k.TallyVotes(ctx, proposal)
+
+// 				// cast vote on host
+// 				k.CastVoteOnHost(ctx, hostZone, vote)	// submits vote ICA on host
+// 			}
+// 		}
+// 	}
+// }
+// function may also want to prune old proposals
```

Comment:
I have a slight bias to move pruning to a separate function, just to make things a bit easier to follow